cmd/list.go

```diff
@@ -0,0 +1,280 @@
+		Category:  "SECRETS",
+		Flags: []cli.Flag{
+			orgFlag("Use this organization.", false),
+			projectFlag("Use this project.", false),
+			envSliceFlag("Use this environment.", false),
+			serviceSliceFlag("Use this service.", "", false),
+			nameFlag("Find secrets with this name."),
+			cli.BoolFlag{
+				Name:  "verbose, v",
+				Usage: "Lists the sources of the secrets (shortcut for --format verbose)",
+			},
+		},
+		Action: chain(
+			ensureDaemon, ensureSession, loadDirPrefs, checkRequiredFlags, listCmd,
+		),
+	}
+	Cmds = append(Cmds, list)
+}
+
+func listCmd(ctx *cli.Context) error {
+	verbose := ctx.Bool("verbose")
+
+	args := ctx.Args()
+
+	cfg, err := config.LoadConfig()
+	if err != nil {
+		return err
+	}
+
+	client := api.NewClient(cfg)
+	c := context.Background()
+
+	// TODO - Make this a type
+	credentialsTree := make(map[string]map[string]credentialSet)
+
+	// Retrieve org and project flag values
+	orgName := ctx.String("org")
+	var org *envelope.Org
+	if orgName == "" {
+		// Retrieve list of available orgs
+		orgs, err := client.Orgs.List(c)
+		if err != nil {
+			return errs.NewExitError("Failed to retrieve orgs list.")
+		}
+
+		// Prompt user to select from list of existing orgs
+		idx, _, err := SelectExistingOrgPrompt(orgs)
+		if err != nil {
+			return errs.NewExitError("Failed to select org.")
+		}
+
+		org = &orgs[idx]
+		orgName = org.Body.Name
+
+	} else {
+		// If org flag was used, identify the org supplied.
+		org, err = client.Orgs.GetByName(c, orgName)
+		if org == nil || err != nil {
+			return errs.NewExitError("org" + orgName + "not found.")
+		}
+	}
+
+	projectName := ctx.String("project")
+	var project *envelope.Project
+	if projectName == "" {
+		// Retrieve list of available projects
+		projects, err := client.Projects.List(c, org.ID)
+		if err != nil {
+			return errs.NewExitError("Failed to retrieve projects list.")
+		}
+
+		// Prompt user to select from list of existing orgs
+		idx, _, err := SelectExistingProjectPrompt(projects)
+		if err != nil {
+			return errs.NewExitError("Failed to select project.")
+		}
+
+		project = &projects[idx]
+		projectName = project.Body.Name
+
+	} else {
+		// Get Project for project name, confirm project exists
+		project, err = getProject(c, client, org.ID, projectName)
+		if err != nil {
+			return errs.NewErrorExitError("Failed to get project info for "+
+				orgName+"/"+projectName, err)
+		}
+	}
+
+	// Retrieve environment flag values
+	// If no values were set, use a full glob
+	envFilters := ctx.StringSlice("environment")
+	if len(envFilters) == 0 {
+		envFilters = append(envFilters, "*")
+	}
+
+	// Retrieve service flag values
+	// If no values were set, use a full glob
+	serviceFilters := ctx.StringSlice("service")
+	if len(serviceFilters) == 0 {
+		serviceFilters = append(serviceFilters, "*")
+	}
+
+	// The following two slices are placeholders necessary to
+	// build the PathExp later. Check that this is true, if I pass
```

Comment:
Did you answer your question you left here?

Reply:
Yep, original code is right and necessary.